src/sipmpdf/functions.py

```diff
@@ -218,31 +219,144 @@ def darkcurrent_response(x: np.float64,
+  """
+  Modified dark current response, as the unmodified dark current response has
+  excessively sharp features that can make other numerical tools unstable, the
+  modified version attempts smooth out the dark current response by substituting
+  out the sharp features at x~pedestal and x~(pedestal+gain) with a normal
+  response with a width of ~0.01 resolution (since we don't expect any detector
+  to have such low a noise, 0.01 feels like a reasonable cutoff)
+
+  Here we modify the function to preserve the overall normalization and the 0-th
+  order smoothness of the function (function is continuous, but derivative is
+  not):
+
+  - If x is in the "nominal region" of (x > pedestal + 0.01 * gain) and (x <
+    pedestal - 0.01 * gain), we will return the original response function.
+
+
+  """
+  kern = kernel_switch(x, gain, resolution)
+  coarse_res = resolution * 50
+  coarse_res = kern.where(coarse_res > 0.01, 0.01, coarse_res)
+  eps = gain * resolution
+  coarse = gain * coarse_res
+
+  nominal = kern.where(((x > coarse) & (x < (gain - coarse))),  #
+                       darkcurrent_response_original(x, gain, resolution), 0.0)
+
+  outer = normal(x, 0, coarse) + normal(x, gain, coarse)
+  outer_norm = (kern.log(coarse_res) - kern.log(resolution)) / 2.0 / (
+    kern.log(1 - resolution) - kern.log(resolution))
+  return nominal + outer * outer_norm
 
 def ap_response_smeared(x: np.float64, smear: np.float64, n_ap: np.int64,
                         beta: np.float64) -> np.float64:
-  kern = kernel_switch(x, smear, n_np, beta)
-  pass
+  kern = kernel_switch(x, smear, n_ap, beta)
 
+  return kern.where(n_ap==1,
+                    kern.exp(-x/beta)*(1+kern.erf(x/(smear*kern.sqrt(2))))/(2*beta),
+                    ap_response(x=x,n_ap=n_ap,beta=beta))  
 
+@expand_shape
 def darkcurrent_response_smeared(x: np.float64,
                                  smear: np.float64,
-                                 pedestal: np.float64,
                                  gain: np.float64,
                                  resolution: np.float64 = 1e-4) -> np.float64:
-  kern = kernel_switch(x, smaer, pedestal, gain, resolution)
-  pass
+  kern = kernel_switch(x, smear, gain, resolution)
+
+  n_samples = 1024
 
+  #extend arrays
+  def extend_arr(x):
+    return kern.repeat(x[kern.newaxis,...], n_samples, axis=0)
 
+  x=extend_arr(x)
+  smear=extend_arr(smear)
+  gain=extend_arr(gain)
+  resolution=extend_arr(resolution)
+  
+  delta = 8 * smear / n_samples  # Getting the integration window
+  idx = kern.local_index(x, axis=0)
+  x_pr = -4 * smear + delta * (idx + 0.5)
+
+  # Given that the computed result will exhibit periodicity within the
+  # integration window, here we compute a simple rolling means with 5 parts
+  return kern.sum(darkcurrent_response(x - x_pr, gain, resolution)*normal(x_pr, 0, smear)*delta,axis=0)
+
+@expand_shape
+def afterpulsing_summation(x: np.float64,
```

Comment:
This function does not make sense to be exposed to the users (after pulse response should not depend on the pedestal value or the gain)

Either: 
- Rename the function to `_afterpulse_summation` (follow the convention that methods is a leading `_` should not be directly used by the users), 
- Refactor the afterpulsing summation to not explicitly require `pedestal` and `gain`